File: day81-90/day81/2641. Cousins in Binary Tree II.php
<?php

class TreeNode {
    public $val;
    public $left;
    public $right;
    
    public function __construct($val = 0, $left = null, $right = null) {
        $this->val = $val;
        $this->left = $left;
        $this->right = $right;
    }
}

class BinaryTree {
    public $root;

    public function __construct($arr) {
        $this->root = $this->buildTree($arr);
    }

    public function buildTree($arr) {
        if (empty($arr)) {
            return null;
        }

        $root = new TreeNode($arr[0]);
        $queue = new SplQueue();
        $queue->enqueue($root);
        $index = 1;

        while (!$queue->isEmpty() && $index < count($arr)) {
            $node = $queue->dequeue();

            
            if ($arr[$index] !== null) {
                $node->left = new TreeNode($arr[$index]);
                $queue->enqueue($node->left);
            }
            $index++;

            if ($index >= count($arr)) {
                break;
            }

            
            if ($arr[$index] !== null) {
                $node->right = new TreeNode($arr[$index]);
                $queue->enqueue($node->right);
            }
            $index++;
        }

        return $root;
    }

    public function printTree() {
        if (!$this->root) {
            return "Empty Tree";
        }

        $queue = new SplQueue();
        $queue->enqueue($this->root);
        $result = [];

        while (!$queue->isEmpty()) {
            $node = $queue->dequeue();
            if ($node) {
                $result[] = $node->val;
                $queue->enqueue($node->left);
                $queue->enqueue($node->right);
            } else {
                $result[] = null;
            }
        }

        
        while (!empty($result) && end($result) === null) {
            array_pop($result);
        }

        return $result;
    }
}

class Solution {
    public function replaceValueInTree($root) {
        if (!$root) {
            return null;
        }

        $queue = new SplQueue();
        $queue->enqueue($root);
        $root->val = 0;  

        while (!$queue->isEmpty()) {
            $levelSum = 0;
            $siblingsMap = [];

            
            $levelSize = $queue->count();
            for ($i = 0; $i < $levelSize; $i++) {
                $node = $queue->dequeue();
                $siblingsMap[] = [$node->left, $node->right];  

                if ($node->left) {
                    $levelSum += $node->left->val;
                    $queue->enqueue($node->left);
                }
                if ($node->right) {
                    $levelSum += $node->right->val;
                    $queue->enqueue($node->right);
                }
            }

            
            foreach ($siblingsMap as [$left, $right]) {
                $siblingSum = 0;
                if ($left) {
                    $siblingSum += $left->val;
                }
                if ($right) {
                    $siblingSum += $right->val;
                }

                if ($left) {
                    $left->val = $levelSum - $siblingSum;  
                }
                if ($right) {
                    $right->val = $levelSum - $siblingSum;  
                }
            }
        }

        return $root;
    }
}


$binaryTree = new BinaryTree([5,4,9,1,10,null,7]);
#$binaryTree->printTree();
$sol=new Solution();
$sol->replaceValueInTree($binaryTree->root);
print_r($binaryTree->printTree());

File: day81-90/day81/2641. Cousins in Binary Tree II.py
from collections import deque
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class BinaryTree:
    def __init__(self, arr):
        self.root = self.build_tree(arr)
    
    def build_tree(self, arr):
        if not arr:
            return None
        
        
        root = TreeNode(arr[0])
        queue = deque([root])
        index = 1
        
        
        while queue and index < len(arr):
            node = queue.popleft()
            
            
            if arr[index] is not None:
                node.left = TreeNode(arr[index])
                queue.append(node.left)
            index += 1
            
            
            if index >= len(arr):
                break
            
            
            if arr[index] is not None:
                node.right = TreeNode(arr[index])
                queue.append(node.right)
            index += 1
        
        return root
    
    
    def print_tree(self):
        if not self.root:
            return "Empty Tree"
        
        queue = deque([self.root])
        result = []
        while queue:
            node = queue.popleft()
            if node:
                result.append(node.val)
                queue.append(node.left)
                queue.append(node.right)
            else:
                result.append(None)
        
        
        while result and result[-1] is None:
            result.pop()
        
        return result
class Solution:
    
    def replaceValueInTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None
        q = deque([root])
        root.val = 0  
        while q:
            level_sum = 0
            siblings_map = {}  
            
            for _ in range(len(q)):
                node = q.popleft()
                siblings_map[node] = (node.left, node.right)  
                if node.left:
                    level_sum += node.left.val
                    q.append(node.left)
                if node.right:
                    level_sum += node.right.val
                    q.append(node.right)
            
            
            for node, (left, right) in siblings_map.items():
                sibling_sum = 0
                if left:
                    sibling_sum += left.val
                if right:
                    sibling_sum += right.val
                
                if left:
                    left.val = level_sum - sibling_sum  
                if right:
                    right.val = level_sum - sibling_sum  
        
        return root
                

arr=[5,4,9,1,10,None,7]
bt=BinaryTree(arr)
root=bt.root        

print(bt.print_tree())

sol=Solution()
root=sol.replaceValueInTree2(root)
print(bt.print_tree())